Optimise more pattenrs

guillep opened the pull request:
Stores and loads, use pattern based optimisations in primitives too

guillep committed:
Optimising more patterns

PalumboN commented on smalltalksrc/VMMaker/DruidJIT.class.st:
```diff
@@ -1469,21 +1472,18 @@ DruidJIT >> gen_PrimitiveAt [
-	self ArithmeticShiftRightCq: 8 R: Extra1Reg.
+	self MoveR: ReceiverResultReg R: Extra0Reg.
+	self SubCq: 8 R: Extra0Reg.
+	self MoveM64: 0 r: Extra0Reg R: Extra0Reg.
+	self LogicalShiftLeftCq: 8 R: Extra0Reg.
+	self ArithmeticShiftRightCq: 8 R: Extra0Reg.
```
It is not optimising for subtraction, but we could put a negative number as offset, right?

PalumboN commented on smalltalksrc/VMMaker/DruidJIT.class.st:
```diff
@@ -2353,112 +2319,109 @@ DruidJIT >> gen_PrimitiveNewWithArg [
+	self MoveMw: 32 r: SPReg R: Extra3Reg.
+	self MoveR: Extra3Reg R: Extra0Reg.
+	self AddR: Extra2Reg R: Extra0Reg.
+	self AddCq: 8 R: Extra0Reg.
+	self SubCq: 1 R: Extra0Reg.
```
We are not optimising adds and subs mixed 🤯

PalumboN commented on smalltalksrc/VMMaker/DruidJIT.class.st:
```diff
@@ -2353,112 +2319,109 @@ DruidJIT >> gen_PrimitiveNewWithArg [
-	self MoveR: Extra3Reg M64: 0 r: SendNumArgsReg.
-	self MoveR: SendNumArgsReg R: ClassReg.
-	self AddCq: 8 R: ClassReg.
+	self MoveR: Extra3Reg M64: 0 r: [ClassReg].
```
There is a block here? 🤔

guillep replied:
Fixed!

PalumboN commented on smalltalksrc/VMMaker/DruidJIT.class.st:
```diff
@@ -2988,14 +2951,69 @@ DruidJIT >> gen_PushLiteralVariable16CasesBytecode0 [
 	jump2 jmpTarget: currentBlock.
-	self AddCq: 8 R: SendNumArgsReg.
-	self ssPushBase: SendNumArgsReg offset: ValueIndex << 3.
+	self ssPushBase: SendNumArgsReg offset: ValueIndex << 3 + 8.
```
Here is there no parenthesis because of the formatter?

guillep replied:
yes, no need (I hope). You are the one that did the folding of jit compiled expressions, you know better than me if they are respecting precedence :D

PalumboN replied:
Yes, I'm asking because it put always the parenthesis. But I know that the formatter removes unnecessary ones. So I hope too :P

PalumboN commented on smalltalksrc/VMMaker/DruidJIT.class.st:
```diff
@@ -3126,8 +3136,7 @@ DruidJIT >> gen_PushReceiverVariableBytecode9 [
 	| currentBlock |
 	self ensureReceiverResultRegContainsSelf.
 	self MoveR: ReceiverResultReg R: TempReg.
-	self AddCq: 8 R: TempReg.
-	self ssPushBase: TempReg offset: 72.
+	self ssPushBase: TempReg offset: 80.
```
🤩

PalumboN reviewed (approved):
Cool! I left some comments. 

I'm not merging it because there is a weird block in the code... Have you built the DruidVM with this code?

guillep committed:
Fix loads
 - 16bit loads only work with aligned offsets for now
 - 32 bit loads should compute their offset correctly

guillep committed:
Fix generated code

guillep commented:
Tests are passing!
